src/rpc/mining.cpp

```diff
@@ -100,6 +100,88 @@ static UniValue getnetworkhashps(const JSONRPCRequest& request)
     return GetNetworkHashPS(!request.params[0].isNull() ? request.params[0].get_int() : 120, !request.params[1].isNull() ? request.params[1].get_int() : -1);
 }
 
+static bool GenerateBlock(const CTxMemPool& mempool, const CScript& coinbase_script, const std::vector<CTransactionRef>* preselected_txs, uint64_t& max_tries, unsigned int& extra_nonce, uint256& block_hash)
+{
+    block_hash.SetNull();
+
+    CBlock block;
+    CChainParams chainparams(Params());
+
+    if (!preselected_txs) {
+        std::unique_ptr<CBlockTemplate> pblocktemplate(BlockAssembler(mempool, chainparams).CreateNewBlock(coinbase_script));
+        if (!pblocktemplate.get())
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Couldn't create new block");
+        block = pblocktemplate->block;
+        {
+            LOCK(cs_main);
+            IncrementExtraNonce(&block, ::ChainActive().Tip(), extra_nonce);
+        }
+    } else {
+
+        CBlockIndex* previous_index;
+        {
+            LOCK(cs_main);
+            previous_index = ::ChainActive().Tip();
+        }
+        CHECK_NONFATAL(previous_index != nullptr);
+
+        const int height = previous_index->nHeight + 1;
+
+        // Create coinbase transaction.
+        CMutableTransaction coinbase_tx;
+        coinbase_tx.vin.resize(1);
+        coinbase_tx.vin[0].prevout.SetNull();
+        coinbase_tx.vout.resize(1);
+        coinbase_tx.vout[0].scriptPubKey = coinbase_script;
+        coinbase_tx.vout[0].nValue = GetBlockSubsidy(height, chainparams.GetConsensus());
+        coinbase_tx.vin[0].scriptSig = CScript() << height << OP_0;
+        block.vtx.push_back(MakeTransactionRef(std::move(coinbase_tx)));
+
+        // Add transactions
+        block.vtx.insert(block.vtx.end(), preselected_txs->begin(), preselected_txs->end());
+
+        block.nVersion = ComputeBlockVersion(previous_index, chainparams.GetConsensus());
+        if (chainparams.MineBlocksOnDemand())
+            block.nVersion = gArgs.GetArg("-blockversion", block.nVersion);
+
+        // Fill in header
+        block.hashPrevBlock = previous_index->GetBlockHash();
+        block.nTime = GetAdjustedTime();
+        UpdateTime(&block, chainparams.GetConsensus(), previous_index);
+        block.nBits = GetNextWorkRequired(previous_index, &block, chainparams.GetConsensus());
+        block.nNonce = 0;
+
+        GenerateCoinbaseCommitment(block, previous_index, chainparams.GetConsensus());
```

Comment:
Why is this code (and the lines above it) in this function? It could be either inlined in the rpc, or if you prefer wrapped into another function that returns a `shared_ptr` of the block, which gets then passed in here.

This makes this function simpler (has one purpose: mining) and has less redundant parameters passed in (like the mempool, which is not needed in this branch at all)

Reply:
Separated the code out.